### ISSUE TITLE
Add support for AWS RDS credential secrets or just encrypted JSON in AWS TriggerAuth

### ISSUE THREAD

**Issue by sneljo1** [kedacore/keda#6312]
### Proposal

There is a option in secretsmanager to store RDS credentials. This are stored in a fixed format, a key/value object with keys `username, password, engine, host, port, dbClusterIdentifier`. It would be nice to be able to 1-1 map these properties to the properties triggers.

AS I currently understand it, the full value of the response will be mapped to the parameter right now.

### Use-Case

Our use-case would be to directly map this to postgres triggers, but generically, this could also serve different use-cases where users have their own JSON encrypted and need it for different triggers.

### Is this a feature you are interested in implementing yourself?

No

**Comment by JorTurFer** [comment]
I thought that there was a feature request to support encrypted json to parse the json and pass a given item from inside the json to the scaler. I think that this is something that definitively we should support as it's an AWS Secrets Manager capability

**Comment by pravarag** [comment]
Hi @JorTurFer , I'd like to take this up if anyone else hasn't started on it yet.

**Comment by Garvit-77** [comment]
@pravarag as there's no activity currently so are you working on this ?
if not @JorTurFer  could you assign me this issue could be raising the pr with my changes by this weekend

**Comment by pravarag** [comment]
> @pravarag as there's no activity currently so are you working on this ? if not @JorTurFer could you assign me this issue could be raising the pr with my changes by this weekend

Hi please go ahead as I've not made any progress yet on this issue.

/unassign

**Comment by Garvit-77** [comment]
@JorTurFer can i please work on this issue .

**Comment by Garvit-77** [comment]
@JorTurFer @sneljo1  
Could you assign me the issue so I could make my push

**Comment by JorTurFer** [comment]
Hello
Sorry for the delay, my life has been complicated, I've assigned it to you! Thanks a lot

**Comment by mpechner** [comment]
Now that keda understands  aws secret keys, maybe easier to implement.
I'd like to take a crack at this.

**Comment by mpechner** [comment]
Is this being worked on.  I can take a shot at it.

Not completely sure what is being asked for. 
Setting up a rds   credential in AWS secret secret manager with scheduled rotation,  I understand.

Writing a postgres trigger to call a go function to do the rotation (thanks chat gpt) Do able.

Unsure of the keda feature being asked for. Especially since RDS creds in AWS secreet manager can be set for automatic rotation.

Add the discussion tag back?

**Comment by JorTurFer** [comment]
I think that nobody is working on this at the moment. I can assign it to you if you are still interested

**Comment by mpechner** [comment]
The more I think about this the more I think other than a helper that returns a string map of a decrypted secret string not sure what the ask is.  Add the helper call to pkg/scaling/resolver/aws_secretmanager_handler.go

Add to AwsSecretManagerhandler
calls read, returns a map[string]string{} or and error If not a json secret either error or map[''][SECRET STRING]

Cannot just change read API since it would require a change to the signature to return map[string]string{} instead of string

func (ash *AwsSecretManagerHandler) ReadJson(ctx context.Context, logger logr.Logger, secretName, versionID, versionStage string) (map[string]string{}, error) {

or call it decodesecret or  just decode?

**Comment by mpechner** [comment]
I found the code that relates to the trigger reference. :-D

I believe I need to add the  AwsSecretManagerJson to the CRD. I have not directly worked on CRDs before.  I pointer would help.
@JorTurFer

**Comment by mpechner** [comment]
I took a very wrong approach.   scaleResolvers.go. resolveAuthRef returns a map[string[]string. so getSecretValues() returning a map[string]string cannot be assigned to a map the result field.

Will try again next weekend.

**Comment by mpechner** [comment]
After looking at the code and because of how resolveAuthRef() in pkg.scaling.resolver.scale_resolvers.go works this cannot be done.  The return wold need to be map[string]map[string]string. instead of map[string]string

Both me and mutliple queries via AI in cursor agree.  

The secretvalue is already able to be returned when triggerAuthSpec.AwsSecretManager, then the caller should decode the json.

The return would need to be map[string]map[string]string. instead of map[string]string

It would be up to the caller to decide if the required fields for RDS are n the secret.

Unless someone has hints on another approach, unassign me.    Doing this would require alot of refactoring, this would require a breaking change.

Again new to both GO and this code base, do please, provide some hints if I an wrong.

**Comment by mpechner** [comment]
For an AWS RDS secret using IAM  for the username, is it expected to be a full arn?

**Comment by mpechner** [comment]
OK, I am just adding a convenience method to decode the json

**Comment by JorTurFer** [comment]
>I believe I need to add the AwsSecretManagerJson to the CRD. I have not directly worked on CRDs before. I pointer would help.

To include this new field in the CRD, you just need to include it in the type file with the proper annotations. I think that adding the property to [this struct](https://github.com/kedacore/keda/blob/acf5d2c102823b12d7bc83bf12d99ab7aa3f0a48/apis/keda/v1alpha1/triggerauthentication_types.go#L354) or maybe to[ this other](https://github.com/kedacore/keda/blob/acf5d2c102823b12d7bc83bf12d99ab7aa3f0a48/apis/keda/v1alpha1/triggerauthentication_types.go#L375), depending on how it works. Then, you just need to execute `make manifests` and the CRDs will be updated automatically based on the changes you've done

**Comment by JorTurFer** [comment]
I'm checking the documentation of this integration more in depth, and I'm not sure if KEDA already supports this. I mean, KEDA supports reading secrets in json format, just setting the `secretKey` in the TriggerAuthentication, checking the command from docs
```bash
aws secretsmanager get-secret-value \
    --secret-id 'arn:aws:secretsmanager:eu-west-1:123456789012:secret:rds!db-033d7456-2c96-450d-9d48-f5de3025e51c-xmJRDx'
```
I think that the secret is stored in a regular json secret, is it? if yes, reading it, it's currently supported

**Comment by mpechner** [comment]
Since the RDS secret is a JSON string, the connivence method I added in pkg/utils is a reasonable compromise for solving this issue.

I do not think the refactor around resolveAuthRef() in pkg.scaling.resolver.scale_resolvers.go works this cannot be done. The return wold need to be map[string]map[string]string. instead of map[string]string is worth it.

**Comment by JorTurFer** [comment]
> Since the RDS secret is a JSON string

That was my last comment about, IDK if the RDS secret is just a simple JSON stored as the typical JSON that you can store as secret. If it is, RDS secrets are currently supported as JSON secrets are already supported

**Comment by mpechner** [comment]
Yes, they are just json key/value. So handled. If you accept the pr, the pkg/util add covers it.  make sure the required fields are present.  The only field it cannot confirm is password since it not not required if the used has access to the DB because of a IM role.

Documentation: https://docs.aws.amazon.com/secretsmanager/latest/userguide/reference_secret_json_structure.html

**Comment by JorTurFer** [comment]
The point is that given the current architecture, I'm not sure if adding a new field is a good idea or just an extra effort for end users. I'd see it if just enabling something in the trigger authentication, the mapping of the properties is done automatically, but even that, I'm not sure if hiding it to users is the right way. WDYT @wozniakjan ?
In any case, sorry @mpechner for the work that you have already done until now, I really thought that RDS wasn't just a secret as JSON but a new API that we needed to support

**Comment by mpechner** [comment]
Considering my involvement in this project is 2 fold. 1 learns a product we are actively using.  2. leaning go.  So not lost work actually.  

I do not think this is a total lose.  https://github.com/kedacore/keda/pull/6807

**Comment by mpechner** [comment]
I assume I need to document these calls.  Where do I add the documentation? The other util functions do not seem to be documented.

**Comment by JorTurFer** [comment]
Do you mean to document the current code or so?

**Comment by mpechner** [comment]
> Do you mean to document the current code or so?

Yes, the code in the PR.  But I could not find where existing code from that directory is documented.
